File: go/problem_75.go
package main

import "sort"

func longestIncreasingSubsequence(nums []int) int {
	tails := []int{}
	for _, v := range nums {
		pos := sort.SearchInts(tails, v)
		if pos == len(tails) {
			tails = append(tails, v)
		} else {
			tails[pos] = v
		}
	}
	return len(tails)
}

func problem75() {
	println("Problem 75:")
	assertEqual(longestIncreasingSubsequence([]int{0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15}), 6)
	assertEqual(longestIncreasingSubsequence([]int{1, 2, 3}), 3)
	assertEqual(longestIncreasingSubsequence([]int{3, 2, 1}), 1)
	println("All Tests Passed")
}

File: rust/problem_75.rs
pub fn longest_increasing_subsequence(nums: &[i64]) -> usize {
    let mut tails: Vec<i64> = vec![];
    for &v in nums {
        let pos = tails.partition_point(|&t| t < v);
        if pos == tails.len() { tails.push(v); } else { tails[pos] = v; }
    }
    tails.len()
}

pub fn problem75() {
    println!("Problem 75:");
    assert_eq!(longest_increasing_subsequence(&[0,8,4,12,2,10,6,14,1,9,5,13,3,11,7,15]), 6);
    assert_eq!(longest_increasing_subsequence(&[1,2,3]), 3);
    assert_eq!(longest_increasing_subsequence(&[3,2,1]), 1);
    println!("All Tests Passed");
}
